### PR TITLE
Expo publish apk

### PR DESCRIPTION
**Pull request by cpirich**
* NOTE: all of this is under our existing `exportExpo` experiment
* Enable new code to generate Android APK app builds using new Expo services
* Update `snack-sdk` to `@code-dot-org/snack-sdk@2.2.0-apkurl-v1` (our privately published version with these changes that are not yet integrated publicly)
* The UI is fairly simplistic for the time being (see screenshot below). A new `Create Android App` button appears after the Try In Expo button succeeds. If that button is pressed, you'll see a spinner for 5-10 minutes while you wait for the APK URL. Once it completes, a new URL will appear that can be sent to friends with Android devices - or can be downloaded and submitted to the Google Play Store.
```
```
screen shot 2019-03-07 at 10 34 55 am

```
```
After the app has built, the bottom of the UI dialog looks like this:

```
```

  screen shot 2019-03-07 at 10 52 38 am

### CODE REVIEW

File: apps/src/constants.js [hunk 1]
```diff
@@ -176,3 +176,6 @@ export const ALPHABET =
   'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
 export const CIPHER =
   'Iq61F8kiaUHPGcsY7DgX4yAu3LwtWhnCmeR5pVrJoKfQZMx0BSdlOjEv2TbN9z';
+
+export const EXPO_SESSION_SECRET =
+  '{"id":"450308ef-67ec-4314-a438-60589b9c0fa2","version":1,"expires_at":2000000000000}';
```
Comment (cpirich) [inline apps/src/constants.js]: NOTE: this is not a valid secret. It is here only for development purposes as a fallback so you can figure out where to swap in your own secret for testing.

File: apps/src/templates/export/expo/App.js.ejs [hunk 2]
```diff
@@ -105,7 +105,7 @@ export default class App extends React.Component {
             originWhitelist={['*']}
             allowFileAccess
             source={{uri: indexUri}}
-            style={styles.webView}
+            style={Platform.OS === 'ios' ? styles.webViewIOS : styles.webViewAndroid}
```
Comment (cpirich) [inline apps/src/templates/export/expo/App.js.ejs:108]: This is a separate fix for the fact that applab apps had a black background on Android only. The HTML page itself doesn't seem to have a color set. That will need to be resolved within applab (perhaps as part of our themes work?)

File: apps/package.json [hunk 3]
```diff
@@ -217,6 +217,7 @@
   },
   "dependencies": {
     "@code-dot-org/js-interpreter": "^1.3.8",
+    "@code-dot-org/snack-sdk": "2.2.0-apkurl-v1",
```
Comment (islemaster) [inline apps/package.json]: Just to clarify, we expect this to converge again? Is there a rough timeline for that?

File: apps/src/code-studio/components/AdvancedShareOptions.jsx [hunk 4]
```diff
@@ -154,20 +159,48 @@ class AdvancedShareOptions extends React.Component {
...
+        generatingExpoApk: false,
+        generatingExpoApkError: null,
+        expoApkUri
+      });
+    } catch (e) {
+      this.setState({
+        generatingExpoApk: false,
+        generatingExpoApkError:
+          'Failed to create Android app. Please try again later.'
+      });
```
Comment (islemaster) [inline apps/src/code-studio/components/AdvancedShareOptions.jsx:200]: Would it be helpful to add some monitoring here and track what sort of errors we see?

File: apps/src/constants.js (hunk 1 above)
Comment (islemaster) [reply apps/src/constants.js]: Can we make that more obvious? Put `fake-fake-fake` in the middle or something?

**Review by islemaster — APPROVED** [review]
I love the progress on this feature. ☎️ 🎲 

Thinking ahead: What does automated testing and monitoring of this feature look like?  Does a plan for that exist anywhere?

File: apps/package.json (hunk 3 above)
Comment (joshlory) [reply apps/package.json]: What changes did we have to make that triggered the fork?

**Comment by joshlory** [comment]
> If that button is pressed, you'll see a spinner for 5-10 minutes while you wait for the APK URL.

@ryansloan What's the long-term UX for this feature? I'm assuming we won't just ask the user to wait. Maybe we notify them when it's done 😄.

File: apps/src/gamelab/Exporter.js [hunk 5]
```diff
@@ -289,7 +290,25 @@ export default {
...
+  async generateExpoApk(snackId, config) {
+    const session = new SnackSession({
+      sessionId: `${getEnvironmentPrefix()}-${project.getCurrentId()}`,
+      name: `project-${project.getCurrentId()}`,
+      sdkVersion: '31.0.0',
+      snackId,
+      user: {
+        sessionSecret: config.expoSession || EXPO_SESSION_SECRET
+      }
+    });
```
Comment (joshlory) [inline apps/src/gamelab/Exporter.js:302]: Can this duplicated code be shared somewhere, maybe StudioApp.js?

File: apps/src/constants.js (hunk 1 above)
Comment (joshlory) [reply apps/src/constants.js]: What does the user experience look like here? I'm not sure I understand when/how the new key gets swapped in for the "real" export flow.

Comment (cpirich) [reply apps/src/constants.js]: It comes in via `appOptions` with chef - but not on local dev systems.

File: apps/package.json (hunk 3 above)
Comment (cpirich) [reply apps/package.json]: They (Expo) have an unmerged PR with these changes. We've published them in our fork.

Comment (cpirich) [reply apps/package.json]: @islemaster convergence should be possible within a week, but we may continue to use our private namespace package to stay ahead of their public facing package while in co-development

File: apps/src/gamelab/Exporter.js (hunk 5 above)
Comment (cpirich) [reply apps/src/gamelab/Exporter.js:302]: We need to refactor `Exporter.js` between gamelab and applab. They are similar in many ways, but also different in some important ways.

**Comment by cpirich** [comment]
@joshlory The proposed UX for APK generation is to open a new page with status/progress in a separate tab, though we need to verify that the proposed experience is reasonable.

File: apps/src/code-studio/components/AdvancedShareOptions.jsx (hunk 4 above)
Comment (cpirich) [reply apps/src/code-studio/components/AdvancedShareOptions.jsx:200]: yes, good idea. Although while under experiment, it's just a few of us who will be doing this.

File: apps/package.json (hunk 3 above)
Comment (islemaster) [reply apps/package.json]: :+1: Awesome, I'm on board with that model, just wanted to make sure it's what we had in mind.  I don't think we've settled on a consistent way to document that vs. hard-fork.

**Pull request merged**